// @alldigitalads/nodejs_modules-twig-frontend/plugins/send-files-plugin.js
'use strict';
const fs = require('fs');
const request = require('request');

class SendFilesPlugin {
    constructor(url, options){
        this.url = url;
        this.options = [];
        Object.assign(this.options, options||[]);
    }

    apply(compiller) {
        let self = this;
        compiller.hooks.done.tap('SendFilesPlugin', (compilation) => {
            Object.keys(compilation.compilation.assets).forEach(key => {
                if(/\.(js|css|s[ac]ss)$/.test(key)){
                    return;
                }

                let uploadOptions = false;
                self.options.some(item => {
                    if(
                        (item.filename && key == item.filename)
                        || (item.pattern && item.pattern.test(key))
                    ){
                        uploadOptions = item;
                        return true;
                    }
                    return false;
                });

                if(uploadOptions !== false){
                    let file = compilation.compilation.assets[key];
                    let FormData = {
                        ruleId: uploadOptions.ruleId,
                        mockFile: {
                            value: file._value ? file._value : Buffer.from(file.source()),
                            options: {
                                contentType: 'application/octet-stream',
                                filename: key
                            }
                        }
                    };
                    request({
                        method: 'POST',
                        uri: self.url,
                        formData: FormData
                    }, (error, response, body) => {
                        if(error){
                            console.error(`Upload file "${key}" fail: `, error);
                        }
                        else if(response.statusCode >= 300){
                            console.error(`Upload file "${key}" fail: Server responsed with code ${response.statusCode}`);
                            console.log(body);
                        }
                    });
                }
            });
        });
    }
}

module.exports = SendFilesPlugin;
