// @alldigitalads/nodejs_modules-twig-extended-loader/lib/twig_router.js
var _ = require('lodash');
var os = require('os');

class Router{
    constructor(options){
        this.rules = options.rules || {};
        this.domain = options.domain || 'http://localhost:8080';
    }

    getPath(name, params){
        params = params || {};
        if(!this.rules[name]){
            console.error(`\x1b[1;31mERROR: Route '${name}' is undefined\x1b[0m`);
            return '';
        }

        var str = this.rules[name].prefix + this.rules[name].path;
        var hash = '';
        if(params['#'] !== undefined){
            hash = '#' + params['#'];
            delete params['#'];
        }

        if(Object.keys(params).length > 1){
            var query_str = '';
            Object.keys(params).forEach(function(key){
                if(key == '_keys'){
                    return;
                }
                if(_.includes(str, '{'+key+'}')){
                    str = str.replace('{'+key+'}', params[key]);
                }
                else if(query_str == ''){
                    query_str += key + '=' + encodeURIComponent(params[key]);
                }
                else{
                    query_str += '&' + key + '=' + encodeURIComponent(params[key]);
                }
            });
            str += query_str ? '?'+query_str : '';
        }
        return str + hash;
    }

    getUrl(name, params){
        var URI = this.getPath(name, params);
        if(URI != ''){
            return this.domain + URI;
        }
        throw new Error('Name "'+name+'" for route not found');
    }

    getAbsoluteUrl(path){
        return this.domain+path;
    }
}

module.exports = Router;
